server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java

```diff
@@ -0,0 +1,259 @@
+        zrw.overwritePersistentData(rootBase, rtm.toJson().getBytes(UTF_8), stat.getVersion());
+      }
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      throw new IllegalStateException(
+          "Could not read root metadata from ZooKeeper due to interrupt", ex);
+    } catch (KeeperException ex) {
+      throw new IllegalStateException(
+          "Could not read root metadata from ZooKeeper because of ZooKeeper exception", ex);
+    }
+  }
+
+  @Override
+  public void upgradeRoot(@NonNull ServerContext context) {
+    log.debug("Upgrade root: upgrading to data version {}", METADATA_FILE_JSON_ENCODING);
+
+    var rootName = Ample.DataLevel.METADATA.metaTable();
+    var filesToConvert = getFileReferences(context, rootName);
+    convertFileReferences(context, rootName, filesToConvert);
+    deleteObsoleteReferences(context, rootName);
+  }
+
+  @Override
+  public void upgradeMetadata(@NonNull ServerContext context) {
+    log.debug("Upgrade metadata: upgrading to data version {}", METADATA_FILE_JSON_ENCODING);
+
+    var metaName = Ample.DataLevel.USER.metaTable();
+    var filesToConvert = getFileReferences(context, metaName);
+    convertFileReferences(context, metaName, filesToConvert);
+    deleteObsoleteReferences(context, metaName);
+  }
+
+  private Map<ComparablePair<KeyExtent,String>,Value>
+      getFileReferences(@NonNull final ServerContext context, @NonNull final String tableName) {
+
+    log.trace("gather file references for table: {}", tableName);
+
+    Map<ComparablePair<KeyExtent,String>,Value> filesToConvert = new TreeMap<>();
+
+    try (Scanner scanner = context.createScanner(tableName, Authorizations.EMPTY)) {
+      scanner.fetchColumnFamily(MetadataSchema.TabletsSection.DataFileColumnFamily.STR_NAME);
+      scanner.forEach((k, v) -> {
+        KeyExtent ke = KeyExtent.fromMetaRow(k.getRow());
+        String file = k.getColumnQualifier().toString();
+        // filter out references that are in the correct format already.
+        if (fileNeedsConversion(file)) {
+          var prev = filesToConvert.put(new ComparablePair<>(ke, file), v);
+          if (prev != null) {
+            throw new IllegalStateException(
+                "upgrade for table: " + tableName + " aborted, would have missed: " + prev);
+          }
+        }
+      });
+    } catch (TableNotFoundException ex) {
+      throw new IllegalStateException("failed to read metadata table for upgrade", ex);
+    }
+    log.debug("Number of files to convert for table: {}, number of files: {}", tableName,
+        filesToConvert.size());
+    return filesToConvert;
+  }
+
+  private void convertFileReferences(final ServerContext context, final String tableName,
+      Map<ComparablePair<KeyExtent,String>,Value> filesToConvert) {
+
+    // not using ample to avoid StoredTabletFile because old file ref is incompatible
+    try (AccumuloClient c = Accumulo.newClient().from(context.getProperties()).build();
```

Comment:
Right now the conversion is happening in two steps where the metadata is first scanned to load the file references into a map and then that is iterated over and updated. One thing I am wondering is if this is a problem if there's a lot of files in the system. I'm not sure what a practical upper bounds would be since depends on a lot of factors but if it's say hundreds of thousands or millions that would be a lot to load at once. We could try and break it up maybe into batches (maybe a tablet at a time or some sort of limit on files at one time and keep track of where you left off). 

I haven't really tried this before also thinking maybe we just modify the files while scanning in the same loop to avoid having to load into a map. The main thing I am not sure of is whether or not that would cause issues with the updates and we'd need to use an isolated scanner but that also buffers the entire row into memory as well. 

@keith-turner any thoughts on this?

Reply:
Changed so that the upgrade file conversion, chop and external compaction removals are done in a single pass through the metadata without creating a map of the file references in e8ae63227647. This reduced the three passes through the metadata that was previous done on first implementation to a single pass and no intermediate storage is used.

Did not use an IsolatedScanner - not sure if that is necessary or not.

Reply:
I like these changes a lot, this will be much more efficient in terms of having to make a single pass and also with memory usage as the batch writer will flush as needed. 

The main thing left would be is if we need to use an `IsolatedScanner` or not to buffer the entire row into memory so the changes in place don't cause issues. The code already handles things if already upgraded so the main thing I'd want to avoid is processing the same columns in the row more than once and I am not sure if that could happen or not.